False positive: C402 Unnecessary generator - rewrite as a dict comprehension.

test.py

```python
s = 'a=1,b=2,c=3'
d = dict(pair.split('=') for pair in s.split(','))
print(d)
```

```
(venv) $ flake8 test.py 
test.py:2:5: C402 Unnecessary generator - rewrite as a dict comprehension.
```

Expected no warnings or errors.

This can't be conveniently rewritten as dict comprehension becuase `pair.split('=')` is an expression, not a simple value. In some cases the expression may be expensive and unwise to execute and index twice.

Discovered by running against the Django code base:

https://github.com/django/django/blob/f3427ccb25806bb062becad0d6a0904930b36635/django/contrib/auth/hashers.py#L381

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -8,6 +8,8 @@ Pending Release
 * New release notes here
 * Don't allow installation with Flake8 3.2.0 which doesn't enable the plugin.
   This bug was fixed in Flake8 3.2.1.
+* Fix C402 false positive where a function call prevents using dict
+  comprehension.
 
 1.2.1 (2016-06-27)
 ------------------

--- a/flake8_comprehensions.py
+++ b/flake8_comprehensions.py
@@ -38,17 +38,28 @@ class ComprehensionChecker(object):
             ):
                 if (
                     isinstance(node.args[0], ast.GeneratorExp) and
-                    node.func.id in ('list', 'set', 'dict')
+                    node.func.id in ('list', 'set')
                 ):
                     msg_key = {
                         'list': 'C400',
                         'set': 'C401',
-                        'dict': 'C402',
                     }[node.func.id]
                     yield (
                         node.lineno,
                         node.col_offset,
                         self.messages[msg_key],
+                        type(self),
+                    )
+
+                elif (
+                    isinstance(node.args[0], ast.GeneratorExp) and
+                    not isinstance(node.args[0].elt, ast.Call) and
+                    node.func.id == 'dict'
+                ):
+                    yield (
+                        node.lineno,
+                        node.col_offset,
+                        self.messages['C402'],
                         type(self),
                     )
 

--- a/tests/test_it.py
+++ b/tests/test_it.py
@@ -118,6 +118,14 @@ def test_C402_pass_1():
     assert errors == []
 
 
+def test_C402_pass_2():
+    errors = run_flake8("""
+        lst = ['a=1', 'b=2', 'c=3']
+        dict(pair.split('=') for pair in lst)
+    """)
+    assert errors == []
+
+
 def test_C402_fail_1():
     errors = run_flake8("""
         foo = dict((x, str(x)) for x in range(10))
@@ -137,6 +145,16 @@ def test_C402_fail_2():
     """)
     assert errors == [
         'example.py:1:10: C402 Unnecessary generator - rewrite as a dict comprehension.',
+    ]
+
+
+def test_C402_fail_3():
+    errors = run_flake8("""
+        lst = [('a', 1), ('b', 2), ('c', 3)]
+        dict(pair for pair in lst if pair[1] % 2 == 0)
+    """)
+    assert errors == [
+        'example.py:2:1: C402 Unnecessary generator - rewrite as a dict comprehension.',
     ]
 
 

